File: 5 Rectangle-Compact/main.py
# Задание 5: Реализцаия Tidy-Rectangle-Compact алгоритма

# Входные данные: встроенный планарный граф G с n вершинами максимальной степени четыре;
# ортогональное представление H из G, такое, что все грани имеют прямоугольную форму.

# Результат: чертеж планарной ортогональной сетки Г из G с ортогональным представлением H и 
# минимальной высотой, шириной, площадью и общей длиной ребёр.

import matplotlib.pyplot as plt
from collections import deque

class Rectangle:
  def __init__(self, id, top, left, bottom, right):
    self.id = id
    self.top = top
    self.left = left
    self.bottom = bottom
    self.right = right
    self.width = right - left
    self.height = top - bottom
    self.drawn = False

class Edge:
  def __init__(self, to, cap, flow, link):
    self.to = to
    self.cap = cap
    self.flow = flow
    self.link = link
    

class Graph:
  def __init__(self, N, S, T):
    self.N = N
    self.S = S
    self.T = T
    self.level = [0 for _ in range(N)]
    self.queue = [0 for _ in range(N)]
    self.head = [-1 for _ in range(N)]
    self.edges = []
    self.stack = deque()

  def AddEdge(self, a, b, cap):
    e1 = Edge(b, cap, 0, self.head[a])
    self.head[a] = len(self.edges)
    self.edges.append(e1)
    
  def AddEdges(self, a, b, cap,):
    e1 = Edge(b, cap, 0, self.head[a])
    e2 = Edge(a, 0, 0, self.head[b])
    self.head[a] = len(self.edges)
    self.edges.append(e1)
    self.head[b] = len(self.edges)
    self.edges.append(e2)

  def BFS(self):
    self.level = [-1 for _ in range(self.N)]
    self.level[self.S] = 0
    q_curr = 0
    q_new = 1
    self.queue[0] = self.S
    while (q_curr < q_new and self.level[self.T] == -1):
      v = self.queue[q_curr]
      q_curr += 1
      id = self.head[v]
      while id != -1:
        e = self.edges[id]
        to = e.to
        id = e.link
        if self.level[to] == -1 and e.flow < e.cap:
          self.level[to] = self.level[v] + 1
          self.queue[q_new] = to
          q_new += 1
          
    return self.level[self.T] != -1
 
  def DFS(self, v, flow):
    if not flow:
      return 0
    if v == self.T:
      return flow

    id = self.head[v]
    while id != -1:
      e = self.edges[id]
      to = e.to
      if self.level[to] != self.level[v] + 1:
        id = e.link
        continue
      pushed = self.DFS(to, min(flow, e.cap - e.flow))
      if pushed:
        self.edges[id].flow += pushed
        self.edges[id ^ 1].flow -= pushed
        return pushed
      id = e.link
    return 0

  def Dinic(self):
    if self.S == self.T:
      print("Incorrect max flow params")
      return -1
    flow = 0
    while self.BFS():
      while pushed := self.DFS(self.S, float('inf')):
        flow += pushed
    return flow

  def ReadStack(self):
    if not self.stack:
      return
    x = self.stack.pop()
    self.ReadStack()
    self.stack.append(x)
    self.edges[x].cap += 1
    return

  def FindAllPathsFromS2T(self, v):
    if v == self.T:
      self.ReadStack()
      self.stack.pop()
      return
  
    id = self.head[v]
    while id != -1:
      e = self.edges[id]
      self.stack.append(id)
      to = e.to
      self.FindAllPathsFromS2T(to)
      id = e.link
  
    if self.stack:
      self.stack.pop()
    return

  def FindMinFlow(self):
    self.FindAllPathsFromS2T(self.S)
    max_flow_graph_vert = Graph(self.N, self.S, self.T)
    from_v = [-1 for _ in range(len(self.edges))]
    for i in range(self.N):
      id = self.head[i]
      while id != -1:
        from_v[id] = i;
        id = self.edges[id].link
  
    for i in range(len(self.edges)):
      max_flow_graph_vert.AddEdges(from_v[i], self.edges[i].to, self.edges[i].cap - 1)
  
    max_flow_graph_vert.Dinic()
  
    for i in range(len(self.edges)):
      self.edges[i].flow = self.edges[i].cap - max_flow_graph_vert.edges[i * 2].flow
    return

  def GetSizesBFS(self, rectangles, dim):
    self.level = [-1 for _ in range(self.N)]
    self.level[self.S] = 0
    q_curr = 0
    q_new = 1
    self.queue[0] = self.S
    while (q_curr < q_new):
      v = self.queue[q_curr]
      q_curr += 1
      id = self.head[v]
      size = 0
      while id != -1:
        e = self.edges[id]
        to = e.to
        id = e.link
        size += e.flow
        if self.level[to] == -1:
          self.level[to] = self.level[v] + 1
          self.queue[q_new] = to
          q_new += 1
        if v != self.S and v != self.T:
          if dim == 'width':
            rectangles[v - 1].width = size
          else:
            rectangles[v - 1].height = size
    return

def PrintResult(filename, rectangles, canvas_border_b, canvas_border_l):
  fig, ax = plt.subplots(nrows=1, ncols=2, figsize=(9, 4.5))
  ax[0].set_title("Source graph")
  for rect in rectangles:
    ax[0].plot([rect.left, rect.right], [rect.top, rect.top], color='0', marker = 'o')
    ax[0].plot([rect.left, rect.right], [rect.bottom, rect.bottom], color='0', marker = 'o')
    ax[0].plot([rect.left, rect.left], [rect.top, rect.bottom], color='0', marker = 'o')
    ax[0].plot([rect.right, rect.right], [rect.top, rect.bottom], color='0', marker = 'o')
  
  new_rectangles = []
  for i in range(len(rectangles)):
    new_rectangles.append(rectangles[i])

  cnt = len(rectangles)
  while cnt > 0:
    for rect_i, rect_candidate in enumerate(rectangles):
      if not rect_candidate.drawn:
        
        left_border = canvas_border_l
        for i, rect_left in enumerate(rectangles):
          if rect_left.drawn and rect_left.right == rect_candidate.left:
            left_border = new_rectangles[i].right
            break
        if left_border == canvas_border_l and rect_candidate.left != canvas_border_l:
          continue
          
        bottom_border = canvas_border_b
        for i, rect_bottom in enumerate(rectangles):
          if rect_bottom.drawn and rect_bottom.top == rect_candidate.bottom:
            bottom_border = new_rectangles[i].top
            break
        if bottom_border == canvas_border_b and rect_candidate.bottom != canvas_border_b:
          continue

        rect_candidate.drawn = True
        cnt -= 1
        new_rectangles[rect_i] = Rectangle(rect_candidate.id, \
                                          bottom_border + rect_candidate.height, \
                                          left_border, \
                                          bottom_border, \
                                          left_border + rect_candidate.width
                                         )

  ax[1].set_title("Result compact graph")
  for rect in new_rectangles:
    ax[1].plot([rect.left, rect.right], [rect.top, rect.top], color='0', marker = 'o')
    ax[1].plot([rect.left, rect.right], [rect.bottom, rect.bottom], color='0', marker = 'o')
    ax[1].plot([rect.left, rect.left], [rect.top, rect.bottom], color='0', marker = 'o')
    ax[1].plot([rect.right, rect.right], [rect.top, rect.bottom], color='0', marker = 'o')
  
  plt.savefig("result_" + filename + ".png")
  plt.show()
  
def TidyRectangleCompact(filename):
  # Rectangles stored in format: id top left bottom right
  rectangles = []
  with open(filename + '.txt') as file:
    lines = file.readlines()
    for line in lines:
      id, top, left, bottom, right = map(float, line.split())
      rect = Rectangle(int(id), top, left, bottom, right)
      rectangles.append(rect)

  s = 0
  n = len(rectangles)
  t = n + 1
  canvas_border_t = max([rect.top for rect in rectangles])
  canvas_border_l = min([rect.left for rect in rectangles])
  canvas_border_b = min([rect.bottom for rect in rectangles])
  canvas_border_r = max([rect.right for rect in rectangles])

  # Build and compute min flow in vertical network
  g_vert = Graph(n + 2, s, t)
  for i in range(n):
    for j in range(i, n):
      if rectangles[i].top == rectangles[j].bottom and \
        not (rectangles[i].left > rectangles[j].right or \
        rectangles[j].left > rectangles[i].right):
        g_vert.AddEdge(rectangles[i].id, rectangles[j].id, 0)
      if rectangles[i].bottom == rectangles[j].top and \
        not (rectangles[i].left > rectangles[j].right or \
        rectangles[j].left > rectangles[i].right):
        g_vert.AddEdge(rectangles[j].id, rectangles[i].id, 0)
    if rectangles[i].top == canvas_border_t:
      g_vert.AddEdge(rectangles[i].id, t, 0)
    if rectangles[i].bottom == canvas_border_b:
      g_vert.AddEdge(s, rectangles[i].id, 0)

  g_vert.FindMinFlow()

  # Build and compute min flow in horizontal network
  g_hor = Graph(n + 2, s, t)
  for i in range(n):
    for j in range(i, n):
      if rectangles[i].right == rectangles[j].left and \
        not (rectangles[i].bottom > rectangles[j].top or \
        rectangles[j].bottom > rectangles[i].top):
        g_hor.AddEdge(rectangles[i].id, rectangles[j].id, 0)
      if rectangles[i].left == rectangles[j].right and \
        not (rectangles[i].bottom > rectangles[j].top or \
        rectangles[j].bottom > rectangles[i].top):
        g_hor.AddEdge(rectangles[j].id, rectangles[i].id, 0)
    if rectangles[i].right == canvas_border_r:
      g_hor.AddEdge(rectangles[i].id, t, 0)
    if rectangles[i].left == canvas_border_l:
      g_hor.AddEdge(s, rectangles[i].id, 0)

  g_hor.FindMinFlow()
  
  # Get new height and width in each rectangle by the minimum flow
  g_hor.GetSizesBFS(rectangles, "height")
  g_vert.GetSizesBFS(rectangles, "width")
  PrintResult(filename, rectangles, canvas_border_b, canvas_border_l)

if __name__ == "__main__":
  TidyRectangleCompact("rectangle_test1")
  TidyRectangleCompact("rectangle_test2")

File: 5 Rectangle-Compact/main.cpp
#include <iostream>
#include <iomanip>
#include <vector>
#include <cstring>

#include<bits/stdc++.h>

const int INF = 1000; // константа-бесконечность

struct Edge {
  int to, cap, flow, link;
};

class Graph
{
public:
    int n, s, t; // number of vertex
    int *level, *queue, *head; // stores level of a node
    std::vector<Edge> edges;
    std::stack<int> stack;
public:
    Graph(int n, int s, int t)
    {
        this->n = n;
        this->s = s;
        this->t = t;
        level = new int[n];
        queue = new int[n];
        head = new int[n];
        memset (head, -1, n * sizeof(int));
    }
 
    // add edge to the graph
    void AddEdges (int a, int b, int cap) {
      Edge e1 = { b, cap, 0, head[a] };
      Edge e2 = { a, 0, 0, head[b] };

      head[a] = (int) edges.size();
      edges.push_back (e1);

      head[b] = (int) edges.size();
      edges.push_back (e2);
    }

    void AddEdge (int a, int b, int cap) {
      Edge e1 = { b, cap, 0, head[a] };

      head[a] = (int) edges.size();
      edges.push_back (e1);
    }
 
    bool BFS();
    int DFS(int v, int flow);
    int Dinic();
    int MinCostFlow();
    int MyDFS(int v);
    int MinFlow();
    void ReadStack();
};

bool Graph::BFS() {
  int q_curr_v = 0, q_new_v = 0;
  queue[q_new_v++] = s;
  memset (level, -1, n * sizeof(level[0]));
  level[s] = 0;
  while (q_curr_v < q_new_v && level[t] == -1) {
    int v = queue[q_curr_v++];
    int id = head[v];
    while (id != -1) {
      Edge e = edges[id];
      int to = e.to;
      id = e.link;
      if (level[to] == -1 && e.flow < e.cap) {
        queue[q_new_v++] = to;
        level[to] = level[v] + 1;
      }
    }
  }
  return level[t] != -1;
}

int Graph::DFS (int v, int flow) {
  if (!flow)  return 0;
  if (v == t)  return flow;

  int id = head[v];
  while (id != -1) {
    // std::cout << "v: " << v << ", id: " << id << std::endl;
    Edge e = edges[id];


    int to = e.to;
    if (level[to] != level[v] + 1) {
      id = e.link;
      continue;
    }
    int pushed = DFS(to, std::min(flow, e.cap - e.flow));
    if (pushed) {
      edges[id].flow += pushed;
      edges[id ^ 1].flow -= pushed;
      // std::cout << "e: " << e.flow << ", e_rev: " << e_rev.flow << ", pushed: " << pushed << std::endl;
      return pushed;
    }
    id = e.link;
  }
  return 0;
}

void Graph::ReadStack()
{
    if(stack.empty())
    {
        std::cout << std::endl;
        return;
    }
    int x = stack.top();
    stack.pop();
    ReadStack();
    stack.push(x);
    std::cout << x << " ";
    edges[x].cap += 1;
    return;
 }

int Graph::MyDFS (int v) {
  if (v == t) {
    ReadStack();
    stack.pop();
    return 0;
  }

  int id = head[v];
  
  while (id != -1) {
    Edge e = edges[id];
    stack.push(id);
    int to = e.to;
    MyDFS(to);
    id = e.link;
  }

  if (v == s) {
    std::cout << std::endl << "Edges: ";
    for (size_t i = 0; i < edges.size(); ++i) {
        std::cout << std::setw(2) << i << " ";
    }
    std::cout << std::endl;
    std::cout << "Capss: ";
    for (size_t i = 0; i < edges.size(); ++i) {
        std::cout << std::setw(2) << edges[i].cap << " ";
    }
    std::cout << std::endl;
  }
  stack.pop();
  return 0;
}

int Graph::Dinic() {
  int flow = 0;
  // std::cout << "Nums: ";
  // for (int i = 0; i < n; ++i) {
  //   std::cout << std::setw(2) << i << " ";
  // }
  // std::cout << std::endl;

  // std::cout << "Head: ";
  // for (int i = 0; i < n; ++i) {
  //   std::cout << std::setw(2) << head[i] << " ";
  // }
  // std::cout << std::endl;


  // std::cout << "Edges: ";
  // for (size_t i = 0; i < edges.size(); ++i) {
  //   std::cout << std::setw(2) << i << " ";
  // }
  // std::cout << std::endl;

  // std::cout << "Capss: ";
  // for (size_t i = 0; i < edges.size(); ++i) {
  //     std::cout << std::setw(2) << edges[i].cap << " ";
  // }
  // std::cout << std::endl;

  // std::cout << "Links: ";
  // for (size_t i = 0; i < edges.size(); ++i) {
  //   std::cout << std::setw(2) << edges[i].link << " ";
  // }
  // std::cout << std::endl;



  while (true) {
    if (!BFS())
      break;
    while (int pushed = DFS(s, INT_MAX)) {
      std::cout << pushed << std::endl;
      flow += pushed;
    }
  }

  std::cout << "Flows Dinic: ";
  for (size_t i = 0; i < edges.size(); ++i) {
    std::cout << std::setw(2) << edges[i].flow << " ";
  }
  std::cout << std::endl;

  return flow;
}

int main (int argc, char **argv) {
  int s = 0;
  int t = 11;
  int n = 12;

  Graph g(n, s, t);
  int a, b;
  while (std::cin >> a >> b) {
    // std::cout << a << " " << b << std::endl;
    g.AddEdge(a, b, 0);
  }

  g.MyDFS(s);

  Graph g_final(n, s, t);

  std::vector<int> from(g.edges.size());
  for (int i = 0; i < n; ++i) {
    int id = g.head[i];
    while (id != -1) {
      from[id] = i;
      id = g.edges[id].link;
    }
  }
  for (size_t i = 0; i < g.edges.size(); ++i) {
    g_final.AddEdges(from[i], g.edges[i].to, g.edges[i].cap - 1);
  }

  int flow = g_final.Dinic();
  std::cout << "Flow: " << flow << std::endl;


  for (size_t i = 0; i < g.edges.size(); ++i) {
    g.edges[i].flow = g.edges[i].cap - g_final.edges[i * 2].flow;
  }

  std::cout << "Flows: ";
  for (size_t i = 0; i < g.edges.size(); ++i) {
    std::cout << std::setw(2) << g.edges[i].flow << " ";
  }
  std::cout << std::endl;
  return 0;
}


















// int Graph::MinCostFlow() {


//   std::cout << "Nums: ";
//   for (int i = 0; i < n; ++i) {
//     std::cout << std::setw(2) << i << " ";
//   }
//   std::cout << std::endl;

//   std::cout << "Head: ";
//   for (int i = 0; i < n; ++i) {
//     std::cout << std::setw(2) << head[i] << " ";
//   }
//   std::cout << std::endl;


//   std::cout << "Edges: ";
//   for (size_t i = 0; i < edges.size(); ++i) {
//     std::cout << std::setw(2) << i << " ";
//   }
//   std::cout << std::endl;

//   std::cout << "Links: ";
//   for (size_t i = 0; i < edges.size(); ++i) {
//     std::cout << std::setw(2) << edges[i].link << " ";
//   }
//   std::cout << std::endl;


//   int flow = 0,  cost = 0;
//   while (flow < k) {
//     std::vector<int> id (n, 0);
//     std::vector<int> d (n, INF);
//     std::vector<int> q (n);
//     std::vector<int> p (n);
//     std::vector<size_t> p_rib (n);
//     int qh=0, qt=0;
//     q[qt++] = s;
//     d[s] = 0;
//     while (qh != qt) {
//       int v = q[qh++];
//       id[v] = 2;
//       if (qh == n)  qh = 0;
//       // for (size_t i=0; i<g[v].size(); ++i) {
//       int ids = head[v];
//       while (ids != -1) {
//         // rib & r = g[v][i];
//         Edge &r = edges[ids];
//         if (r.flow < r.cap && d[v] + r.cost < d[r.to]) {
//           d[r.to] = d[v] + r.cost;
//           if (id[r.to] == 0) {
//             q[qt++] = r.to;
//             if (qt == n)  qt = 0;
//           }
//           else if (id[r.to] == 2) {
//             if (--qh == -1)  qh = n-1;
//             q[qh] = r.to;
//           }
//           id[r.to] = 1;
//           p[r.to] = v;
//           p_rib[r.to] = ids;
//         }
//         ids = r.link;
//       }
//     }

//     // std::cout << "Dist: ";
//     // for (int i = 0; i < n; ++i) {
//     //   std::cout << std::setw(2) << d[i] << " ";
//     // }
//     // std::cout << std::endl;

//     if (d[t] == INF) {
//       std::cout << "here" << std::endl;
//       break;
//     }
//     int addflow = k - flow;
//     for (int v = t; v != s; v = p[v]) {
//       int pv = p[v];
//       size_t pr = p_rib[v];

//       size_t ids = head[pv];
//       Edge es = edges[ids];
//       while (ids != pr) {
//         ids = es.link;
//         es = edges[ids];
//       }
//       // addflow = min(addflow, g[pv][pr].u - g[pv][pr].f);
//       addflow = std::min(addflow, es.cap - es.flow);
//     }
//     for (int v = t; v != s; v = p[v]) {
//       int pv = p[v];
//       size_t pr = p_rib[v];

//       size_t r = pr + 1;
//       // size_t r = g[pv][pr].back;

//       size_t ids = head[pv];
//       Edge es = edges[ids];
//       while (ids != pr) {
//         ids = es.link;
//         es = edges[ids];
//       }

//       edges[ids].flow += addflow;
//       cost += edges[ids].cost * addflow;
      
//       ids = head[v];
//       es = edges[ids];
//       while (ids != r) {
//         ids = es.link;
//         es = edges[ids];
//       }

//       edges[ids].flow -= addflow;


//       // g[pv][pr].f += addflow;
//       // g[v][r].f -= addflow;
//       // cost += g[pv][pr].c * addflow;
//     }
//     flow += addflow;
//   }

//   std::cout << "Flows: ";
//   for (size_t i = 0; i < edges.size(); ++i) {
//     std::cout << std::setw(2) << edges[i].flow << " ";
//   }
//   std::cout << std::endl;

//   std::cout << "Cost: " << cost << std::endl;
//   return flow;
// }